neuralpp/symbolic/sympy_expression.py

```diff
@@ -1,75 +1,121 @@
 # and SymPyExpression. I usually use "sympy object" to refer to the former and "expression" to refer to the latter.
 
 
-def python_callable_to_sympy_function(python_callable: Callable) -> Type[sympy.Basic]:
-    match python_callable:
-        # boolean operation
-        case operator.__and__:
-            return sympy.And
-        case operator.__or__:
-            return sympy.Or
-        case operator.__not__:
-            return sympy.Not
-        case operator.__xor__:
-            return sympy.Xor
-        # comparison
-        case operator.le:
-            return sympy.Le
-        case operator.lt:
-            return sympy.Lt
-        case operator.ge:
-            return sympy.Ge
-        case operator.gt:
-            return sympy.Gt
-        case operator.eq:
-            return sympy.Eq
-        # arithmetic
-        case operator.add:
-            return sympy.Add
-        case operator.mul:
-            return sympy.Mul
-        case operator.pow:
-            return sympy.Pow
-        # min/max
-        case builtins.min:
-            return sympy.Min
-        case builtins.max:
-            return sympy.Max
+def infer_sympy_object_type(sympy_object: sympy.Basic, type_dict: Dict[sympy.Basic, Expression]) -> Expression:
+    match sympy_object:
+        case sympy.Integer():
+            return BasicConstant(int)
+        case sympy.Float():
+            return BasicConstant(float)
+        case sympy.Rational():
+            return BasicConstant(fractions.Fraction)
+        case sympy.logic.boolalg.BooleanAtom():
+            return BasicConstant(bool)
         case _:
-            raise ValueError(f"Python callable {python_callable} is not recognized.")
+            # It is obvious that we can look up type_dict for symbols like `symbol("x")`.
+            # We can also look up for function, such as `a+b`, where type_dict records the function type (instead of
+            # return type) of the (root-level) function.
+            try:
+                return type_dict[sympy_object]
+            except KeyError:  # if it's not in type_dict, try figure out ourselves
+                return BasicConstant(infer_python_callable_type(sympy_function_to_python_callable(sympy_object)))
+
+
+python_callable_and_sympy_function_pair = [
+    # boolean operation
+    (operator.__and__, sympy.And),
+    (operator.__or__, sympy.Or),
+    (operator.__not__, sympy.Not),
+    (operator.__xor__, sympy.Xor),
+    # comparison
+    (operator.le, sympy.Le),
+    (operator.lt, sympy.Lt),
+    (operator.ge, sympy.Ge),
+    (operator.gt, sympy.Gt),
+    (operator.eq, sympy.Eq),
+    # arithmetic
+    (operator.add, sympy.Add),
+    (operator.mul, sympy.Mul),
+    (operator.pow, sympy.Pow),
+    # min/max
+    (builtins.min, sympy.Min),
+    (builtins.max, sympy.Max),
+]
+
+
+def sympy_function_to_python_callable(sympy_function: sympy.Basic) -> Callable:
+    try:
+        return {sympy_function: python_callable
```

Comment:
Here you are building this dict every time you look a function up. It seems better to make the dict a property created only the first time it is requested.

Reply:
I just created a global one... I don't want to go through the "emulating a static variable of a function" stuff...